### ISSUE TITLE
Approved form does not become a ticket [V2.12.1]

### ISSUE THREAD

**Issue by phamorais** [pluginsGLPI/formcreator#2366]
**Describe the bug**
After updating from version 2.12.1, when I have a form that needs to be approved, when I approve it it doesn't become a ticket.

Select the ID 333229 to aprove
![image](https://user-images.githubusercontent.com/35368369/130107852-d005d5ce-535d-4c08-9ddb-ebfd3029a481.png)

When I click on "Accept", he still in list 

![image](https://user-images.githubusercontent.com/35368369/130107982-4edb6c58-f4b3-4e09-b253-9b0c070450f9.png)

![image](https://user-images.githubusercontent.com/35368369/130108172-16e8bef5-82fd-4849-849e-ed355d12de89.png)

No errors in php-erros.log

GLPI 9.5.5

**Comment by btry** [comment]
Hi

I had 2 reports  with version 2.12.0, and (one solved, the other waiting for feedback). I think this problem was solved by a bug I found myself a few weeks ago.

There is maybe a correlation with https://github.com/pluginsGLPI/formcreator/issues/2366 .

Please apply the patch found in this comment: https://github.com/pluginsGLPI/formcreator/issues/2349#issuecomment-897645835

It tries to add data to php-errors.log when failing to create a ticket.

**Comment by phamorais** [comment]
@btry Hi!

I apply the patch!

![image](https://user-images.githubusercontent.com/35368369/130135521-431c0508-9d46-46ca-a8c1-549e06d7077f.png)

And I try again in GLPI, but no sucess.

php-errro.log (I don't think it's related to the formcreator)

![image](https://user-images.githubusercontent.com/35368369/130135607-e4cf65cc-9476-43dd-a642-46f729f99538.png)

**Comment by btry** [comment]
Hi

The errors you found are not repated to Formcreator. It seems that your Fields plugin is broken as the missing class name seems to belong to the core of this plugin.

Your result is similar to the other person : no message caught. You can remote the patch.

Let's assume the problem is depends on how the form is built. Please export it, rename it as .txt file and put it here. Be careful if you have confidential data in the form. You may workaround by copying it, removing confidential data, then check the problem persists with this copy and finally export it.

**Comment by phamorais** [comment]
@btry Hi!

Then I realized something.

In the production environment that is with form 2.8 I managed to export the form without problems.

However, when trying to export the same form with version 2.12 I get the error below

Text: "It is not possible to export an empty object: Question regular expression", However, in this form, there is no question with the "Regular expression" field filled in

![image](https://user-images.githubusercontent.com/35368369/130249831-765601a4-cd31-478e-9bb3-189f2b08b9db.png)

Error in php-errors.log

![image](https://user-images.githubusercontent.com/35368369/130250113-60bb00b4-0a8d-4751-861f-cc68cc90cf2b.png)

Follow json from the file in production
[export_formcreator_20210820_1131.txt](https://github.com/pluginsGLPI/formcreator/files/7022273/export_formcreator_20210820_1131.txt)

**Comment by btry** [comment]
Hi

I cannot import a version made with versioin 2.8. This is too old, and I added in 2.9 or maybe earlier a system to avoid import of forms from different versions. 

You must fix your export problem. This is very likely caused by an old bug which where Formcreator "forgot" to create data for regular expressions when a question need it. It does not impacts the forms, only the export. The bug is fixed now, but data still is missing in the DB.

In the impacted form, edit and save all questions able to handle regular expressions. No need to do any change. This should create the missing data; then you should  be able to export the form.

**Comment by phamorais** [comment]
@btry Hi!

Understand.

I took this form as a reference, and it has 27 questions. I did what you recommended, and opened all the questions and saved again, but the problem for exporting remained.

But I noticed a new situation.

It has a form that does not go through the 'approval' process.  That form I manage to clone it, export it without problems.
If I fill in this form, I get approval and it takes a ticket as expected.

![image](https://user-images.githubusercontent.com/35368369/130259341-773ea4a8-afaa-44d9-aa0a-a9d1664e510e.png)

So to test the approval functionality, I took this form and added it to the validation flow.

![image](https://user-images.githubusercontent.com/35368369/130259693-0d3e769f-473d-496f-b541-f2d250271731.png)

After filling it out, it went to the approval queue correctly.

![image](https://user-images.githubusercontent.com/35368369/130259886-7b005f3c-1f78-48e5-a1e0-d8920e4ded94.png)

I click in "Accept", and the form became a ticket As expected.

![image](https://user-images.githubusercontent.com/35368369/130259966-084404fd-c61e-4be4-bc05-a5ec49f51dc7.png)

![image](https://user-images.githubusercontent.com/35368369/130260037-441703f8-d45b-40c7-b03d-96d1ec51d0e0.png)


So to sum up, I believe that some problem in the form causes the 'form approval' functionality to have the bug.

Is there any other way to export a form's json other than through the web interface? Or some specific information in the database regarding this problem form, which we can provide you?

**Comment by phamorais** [comment]
@btry 

Complementing the previous answer, after editing all the questions and saving them, I get this in error.log
![image](https://user-images.githubusercontent.com/35368369/130260488-c685f0f4-f872-4942-a4c0-8845dcc5688a.png)

**Comment by btry** [comment]
Hi

These errors don't generate backtrace. Ad the following patch to GLPI then reproduce the errors on getFromDBByCrit(). You should find a backtrace in php-errors.log. This is needed to find where is the inappropriate cal

```diff
diff --git a/inc/commondbtm.class.php b/inc/commondbtm.class.php
index d2cc10e403..9afeda839d 100644
--- a/inc/commondbtm.class.php
+++ b/inc/commondbtm.class.php
@@ -393,6 +393,7 @@ class CommonDBTM extends CommonGLPI {
                      count($iterator)
                      )
                );
+         Toolbox::backtrace();
       }
       return false;
    }
```

**Comment by phamorais** [comment]
@btry Hi!

I applied the patch.

When trying to approve the form, this was presented in php-errors.log

![image](https://user-images.githubusercontent.com/35368369/130265231-b94848c5-61e8-44e2-84e6-79a9edc64ea2.png)

When I Try export de form, I get this

![image](https://user-images.githubusercontent.com/35368369/130265304-86b9b8c6-6b08-4a91-98cd-219da74d5497.png)

**Comment by btry** [comment]
I searched why no backtraces wed added. I went too fast and the diff adds the backtrace in the wrong method...

Remove the previous patch, add this one instead, And try again. 


````diff
diff --git a/inc/commondbtm.class.php b/inc/commondbtm.class.php
index d2cc10e403..a8597b01aa 100644
--- a/inc/commondbtm.class.php
+++ b/inc/commondbtm.class.php
@@ -351,6 +351,7 @@ class CommonDBTM extends CommonGLPI {
                count($iter)
             )
          );
+         Toolbox::backtrace();
       }
       return false;
    }
```

**Comment by phamorais** [comment]
@btry Hi!

I applied the new patch.
When approving the ticket again, no errors were presented. And when trying to export the form I got the error below

![image](https://user-images.githubusercontent.com/35368369/130269865-f9c00b87-1375-402d-957e-8a94dbc71c9e.png)

**Comment by phamorais** [comment]
Hi @btry 
Did the test above help you with anything?

**Comment by btry** [comment]
Hi

Not yet. It seems that there are duplicate rows in the tables containing min/max values or regex values.

The plugin probably lacks an unicity constraint on these tables.

Please check the tables of the plugin ending with ranges and regexes. They have a foreign key to questions table. Check that the questions having regex and min/max properties have only one row on these tables linked to them. I guess you will find several ones. The count should match the numbers found in the log. Removing the extra rows should help.

**Comment by phamorais** [comment]
@btry Hi!

I was a little confused by the table structure.

I couldn't identify the relationship between the form "glpi_plugin_formcreator_forms" and the question. I thought it would be some column in the table "glpi_plugin_formcreator_questions", but I couldn't find a compatible column.

In the table "glpi_plugin_formcreator_questionregexes" has values filled in, and from what I analyzed the column "plugin_formcreator_questions_id" and the relationship between the answer value and the question id.

So, I went to the table "glpi_plugin_formcreator_questions" but in the columns, I couldn't identify the relationship with the form.


Can u help ?

**Comment by btry** [comment]
Forms contains sections, and sections contains questions. That's why there is no direct relation in DB between questions and forms.

When you view the questions while designing the form, you can move the mouse over the name of the question. The hyperlink is a call to some JS function, with one of the arguments the ID of the question. This will probably be faster to find in the DB the question you need to inspect.

**Comment by phamorais** [comment]
@btry I understand your explanation!

So, returning to the initial problem of this issue, you asked me to export the form for your analysis. When trying to export I get the message mentioned above

![image](https://user-images.githubusercontent.com/35368369/130619738-acce1a36-f7ea-469b-abe5-2af8647bee05.png)

After your explanation, I was able to get to the question values, and see that the "regex" column is empty in all questions na tabela "glpi_plugin_formcreator_questionregexes"

![image](https://user-images.githubusercontent.com/35368369/130619958-e0cd56f1-dbe2-4f76-80fe-85a531780e17.png)

And in table "glpi_plugin_formcreator_questionranges"

![image](https://user-images.githubusercontent.com/35368369/130620153-d9b0afce-a19a-4aee-8d52-771b88d714c7.png)

So I believe the form export bug is not related to REGEX

**Comment by btry** [comment]
When Formcreator exports the questions or finds several rows of regexes or min/max for a single question. This is not normal. Only one row must exists, and for some reason, several were created.

Example there are approximatively 10 rows in the regex table for the question 1392. Delete all, except the first.

**Comment by phamorais** [comment]
@btry Do I need to delete from 2 tables?

**Comment by phamorais** [comment]
@btry Hi!

After a lot of effort I managed to clean the records as you informed.

Question: Any ideas why these records were duplicated? It doesn't make much sense from what you explained. I believe that I will have to make a script to be able to catch all the questions that are duplicated.

Finally,

I had to clean the duplicate records from the tables "glpi_plugin_formcreator_questionranges" and "glpi_plugin_formcreator_questionregexes" to be able to export the .json.

PS: After deleting the duplicate records, I redid the form filling operation, the bug remained, and the form did not become a ticket.

Follow my Form, as requested 

[export_formcreator_20210824_1456.txt]
(https://github.com/pluginsGLPI/formcreator/files/7041254/export_formcreator_20210824_1456.txt)

**Comment by phamorais** [comment]
@btry Hi!

Is there any way to get paid support for this issue, without having to hire a subscription? She is very important to me.

**Comment by btry** [comment]
Hi

Fill the following form

http://portal.glpi-network.com/contact-us

**Comment by phamorais** [comment]
@btry Hi!
I consulted and spoke with Polina, but for her explanation it will not be possible to get paid support under my conditions. I will await your correction. :(

**Comment by btry** [comment]
I'm busy with several tasks unrelated with Formcreator; this may take a while before I can investigate on your issue.

**Comment by phamorais** [comment]
@btry Hi!

I identified that the form I sent you above has a question like "IP". If I remove the question, the error does not occur, so I think it is related to the issue https://github.com/pluginsGLPI/formcreator/issues/2392

**Comment by btry** [comment]
Hi

I did not expect such behavior; then there is a real problem with this type of question. I mark your issue as duplicate of #2392.

**Comment by phamorais** [comment]
@btry hi!

I don't know which branch I should do the PR for you, could you help me?

In my environment, we found this to need improvement. After that, the bug was fixed.

File: inc/formanswer.class.php

![image](https://user-images.githubusercontent.com/35368369/132792534-ce138fce-d0fd-4cd4-acd1-91cead237a70.png)

**Comment by btry** [comment]
Hi

Make your pr based onto support/2.12.0

**Comment by btry** [comment]
By the way, this issue is very likely related to the question of type IP, and your PR after a quick reading, seems not related. When you submit your PR, please describe the bug it shall fix.

**Comment by phamorais** [comment]
@btry Hi!
work for me with https://github.com/pluginsGLPI/formcreator/pull/2398/files

**Comment by btry** [comment]
Hi

Thank you, I'll add this to the next release.

**Comment by phamorais** [comment]
Hi! @btry 
I closed this issue after the tests we ran as it was fixed with version 2.12.2 today.
